pandas/_libs/src/klib/khash_python.h

```diff
@@ -330,6 +333,129 @@ Py_hash_t PANDAS_INLINE tupleobject_hash(PyTupleObject* key) {
+    }
+    else if (unit == NPY_FR_M) {
+        if ((value % 12) == 0) {
+            return hash_datetime_value_and_reso(value / 12, NPY_FR_Y, dts);
+        }
+        return value;
+    }
+    else if (unit == NPY_FR_W) {
+        if (dts->day == 1) {
+            value = (dts->year - 1970) * 12 + dts->month;
+            return hash_datetime_value_and_reso(value, NPY_FR_M, dts);
+        }
+        return value;
+    }
+    else if (unit == NPY_FR_D) {
+        if ((value % 7) == 0) {
+            return hash_datetime_value_and_reso(value / 7, NPY_FR_W, dts);
+        }
+        return value;
+    }
+    else if (unit == NPY_FR_h) {
+        if ((value % 24) == 0) {
+            return hash_datetime_value_and_reso(value / 24, NPY_FR_D, dts);
+        }
+        return value;
+    }
+    else if (unit == NPY_FR_m) {
+        if ((value % 60) == 0) {
+            return hash_datetime_value_and_reso(value / 60, NPY_FR_h, dts);
+        }
+        return value;
+    }
+    else if (unit == NPY_FR_s) {
+        if ((value % 60) == 0) {
+            return hash_datetime_value_and_reso(value / 60, NPY_FR_m, dts);
+        }
+        return value;
+    }
+    else if (unit == NPY_FR_ms) {
+        if ((value % 1000) == 0) {
+            return hash_datetime_value_and_reso(value / 1000, NPY_FR_s, dts);
+        }
+        return value;
+    }
+    else if (unit == NPY_FR_us) {
+        if ((value % 1000) == 0) {
+            return hash_datetime_value_and_reso(value / 1000, NPY_FR_ns, dts);
+        }
+        return value;
+    }
+    else if (unit == NPY_FR_ns) {
+        if ((value % 1000) == 0) {
+            return hash_datetime_value_and_reso(value / 1000, NPY_FR_us, dts);
+        }
+        return value;
+    }
+    else if (unit == NPY_FR_ps) {
+        if ((value % 1000) == 0) {
+            return hash_datetime_value_and_reso(value / 1000, NPY_FR_ns, dts);
+        }
+        return value;
+    }
+    else if (unit == NPY_FR_fs) {
+        if ((value % 1000) == 0) {
+            return hash_datetime_value_and_reso(value / 1000, NPY_FR_ps, dts);
+        }
+        return value;
+    }
+    else if (unit == NPY_FR_as) {
+        if ((value % 1000) == 0) {
+            return hash_datetime_value_and_reso(value / 1000, NPY_FR_fs, dts);
+        }
+        return value;
+    }
+    else {
+        // i.e. NPY_FR_GENERIC
+        // we default to treating these like nanos
+        return hash_datetime_value_and_reso(value, NPY_FR_ns, dts);
+    }
+}
+
+
+// TODO: same thing for timedelta64 objects
+Py_hash_t np_datetime64_object_hash(PyObject* key) {
+    // GH#50690 numpy's hash implementation does not preserve comparabity
+    // either across resolutions or with standard library objects.
+    // See also Timestamp.__hash__
+
+    NPY_DATETIMEUNIT unit = (NPY_DATETIMEUNIT)((PyDatetimeScalarObject*)key)->obmeta.base;
+    npy_datetime value = ((PyDatetimeScalarObject*)key)->obval;
+    npy_datetimestruct dts;
+    PyObject* dt;
```

Comment:
Just a note: You could move this below if you like (same line as the call), we have C99 now :).  (since it is only used locally in the if anyway)

Reply:
sure.  is that just a code style thing or does it affect something like memory allocation in cases where it isnt reached?

Reply:
Just style, just for a long time we used C89 which enforces definitions coming early.